Add support for protobufs in puree-android.

dalewis opened the pull request:
This change is fairly large and opinionated. Puree Android previously worked
entirely with JSON via GSON's JsonObject and family. In order to add generic
protobuf support, I essentially split Puree in two so that it can accept
either JSON log entries OR protobuf log entries, even in the same application.
When registering a log class with PureeConfiguration, you need to specify how
the log should be stored. Log classes which extend PureeLog will use the JSON
logging pathway, and classes which extend MessageLite (the lite version of
protobuf) will automatically use the Protobuf logging pathway.

At the storage level, I added a "format" column to the table which keeps track
of whether a log entry is JSON or protobuf. To simplify things, for now protobuf
is serialized to bytes and then encoded as a base64 string for storage in the
same table. We can probably split into multiple tables in the future, but this
is the simplest solution and is fast enough for our needs since buffered logs
get sent on background threads anyway.

In order to support differences between databases, I simply changed the database
name itself that's used for the sqlite database rather than changing version numbers.
The reason is the following: It's possible that an app might use the original version
of Puree, then upgrade to this one which has a different database format. The upgrade
could work properly with onUpgrade and incrementing the version number, but if for
some reason there was a rollback to a version of Puree from the main repository
rather than this fork, then you could potentially find yourself in a situation where
there was an incompatibility because onDowngrade is not implemented in the main repo.
A simple solution is just changing the database name itself, because in the event of
a rollback to the version in the main repo, that version uses a different database
name and onDowngrade would never be fired.

In order to support protobuf in a generic way, the storage system simply stores the
serialized bytes from the protobuf MessageLite instance, but does not store the class
itself, so the storage system cannot repopulate the object. It's up to the output
plugins to know what type is stored with each log entry. See OutBufferedProtobufLogcat.java
for an example of repopulating a protobuf from the buffer and then wrapping it in another
protobuf.

I upgraded the demo to show how to use protobufs for logging. The protobuf itself
is defined in demo/src/main/proto/example.proto and is a bare minimum protobuf.
The demo also shows how to properly set up gradle on apps in order to compile
protobufs for usage in the app, if that's your thing.

I also fixed a bunch of bugs I found along the way.

TESTED=original unit tests and demo.

darren-mercari committed:
Add support for protobufs in puree-android.

This change is fairly large and opinionated. Puree Android previously worked
entirely with JSON via GSON's JsonObject and family. In order to add generic
protobuf support, I essentially split Puree in two so that it can accept
either JSON log entries OR protobuf log entries, even in the same application.
When registering a log class with PureeConfiguration, you need to specify how
the log should be stored. Log classes which extend PureeLog will use the JSON
logging pathway, and classes which extend MessageLite (the lite version of
protobuf) will automatically use the Protobuf logging pathway.

At the storage level, I added a "format" column to the table which keeps track
of whether a log entry is JSON or protobuf. To simplify things, for now protobuf
is serialized to bytes and then encoded as a base64 string for storage in the
same table. We can probably split into multiple tables in the future, but this
is the simplest solution and is fast enough for our needs since buffered logs
get sent on background threads anyway.

In order to support differences between databases, I simply changed the database
name itself that's used for the sqlite database rather than changing version numbers.
The reason is the following: It's possible that an app might use the original version
of Puree, then upgrade to this one which has a different database format. The upgrade
could work properly with onUpgrade and incrementing the version number, but if for
some reason there was a rollback to a version of Puree from the main repository
rather than this fork, then you could potentially find yourself in a situation where
there was an incompatibility because onDowngrade is not implemented in the main repo.
A simple solution is just changing the database name itself, because in the event of
a rollback to the version in the main repo, that version uses a different database
name and onDowngrade would never be fired.

In order to support protobuf in a generic way, the storage system simply stores the
serialized bytes from the protobuf MessageLite instance, but does not store the class
itself, so the storage system cannot repopulate the object. It's up to the output
plugins to know what type is stored with each log entry. See OutBufferedProtobufLogcat.java
for an example of repopulating a protobuf from the buffer and then wrapping it in another
protobuf.

I upgraded the demo to show how to use protobufs for logging. The protobuf itself
is defined in demo/src/main/proto/example.proto and is a bare minimum protobuf.
The demo also shows how to properly set up gradle on apps in order to compile
protobufs for usage in the app, if that's your thing.

I also fixed a bunch of bugs I found along the way.

TESTED=original unit tests and demo.

darren-mercari committed:
Remove unused imports.

darren-mercari committed:
Update README so that it reflects protobuf support.

darren-mercari committed:
Add Javadoc to fix warning.

darren-mercari committed:
Change Gradle build settings so we can publish artifacts.

TESTED=Successfully published an artifact to bintray (then deleted).

darren-mercari committed:
Update artifact version to match README.

darren-mercari committed:
Fix unittests.

TESTED=Ran all unittests in Android Studio.

darren-mercari committed:
Modifying build.gradle for jitpack support.

darren-mercari committed:
Try lint options to get jitpack working.

darren-mercari committed:
Tell Gradle where to find the classes to put in the jar.

darren-mercari committed:
Change JAR generation again.

TESTED=local jar looks good (./gradlew androidJar)

darren-mercari committed:
Exclude R and BuildConfig from the jar.

TESTED=./gradlew androidJar && jar tf <path/to/jar>

darren-mercari committed:
Lint options weren't the problem.

dalewis committed:
Merge pull request #2 from mercari/enhancement/add-jitpack-support

Enhancement/add jitpack support

rkowase commented on README.md:
````diff
@@ -161,23 +169,25 @@ public class OutFakeApi extends PureeBufferedOutput {
 
 ```java
-public class AddEventTimeFilter implements PureeFilter {
+public class AddEventTimeFilter implements PureeJsonFilter {
````
It looks breaking change. Should we remain original name for developers using this Library?

dalewis replied:
It's definitely a breaking change. I can switch all the *Json* classes back to their original names, although it might be confusing for new users. Good point, though. This would make migration easier.

rkowase replied:
IMO, we should write a migration guide document from 4.x to 5.x if we make breaking change.

dalewis replied:
@rkowase I updated the README in this pull request with descriptions and explanation. I don't think we need to write a migration guide since we don't even know who uses this library.

dalewis replied:
I renamed those classes back to their original names.

darren-mercari committed:
Change *Json* classes back to their original names for better backwards compatibility.

This makes the new version more of a drop-in replacement for the previous version.

TESTED=unittests, demo

darren-mercari committed:
Merge branch 'epic/add-protobuf-support' of github.com:mercari/puree-android into epic/add-protobuf-support
